## Import Library, mengambil, dan membaca data

In [334]:
import pandas as pd
import matplotlib.pyplot as plt


pd.options.mode.chained_assignment = None #mengabaikan saran dari py
pd.options.display.max_columns = 40 #memaksimalkan tampilan kolom data sampai 50
data = pd.read_csv("Levels_Fyi_Salary_Data.csv")
data

timestamp     company     level                         title  \
0       6/7/2017 11:33:27      Oracle        L3               Product Manager   
1      6/10/2017 17:11:29        eBay      SE 2             Software Engineer   
2      6/11/2017 14:53:57      Amazon        L7               Product Manager   
3       6/17/2017 0:23:14       Apple        M1  Software Engineering Manager   
4      6/20/2017 10:58:51   Microsoft        60             Software Engineer   
...                   ...         ...       ...                           ...   
62637   9/9/2018 11:52:32      Google        T4             Software Engineer   
62638   9/13/2018 8:23:32   Microsoft        62             Software Engineer   
62639  9/13/2018 14:35:59        MSFT        63             Software Engineer   
62640  9/16/2018 16:10:35  Salesforce  Lead MTS             Software Engineer   
62641   1/29/2019 5:12:59       apple      ict3             Software Engineer   

       totalyearlycompensation           location  yearsofexperience  \
0                       127000   Redwood City, CA                1.5   
1                       100000  San Francisco, CA                5.0   
2                       310000        Seattle, WA                8.0   
3                       372000      Sunnyvale, CA                7.0   
4                       157000  Mountain View, CA                5.0   
...                        ...                ...                ...   
62637                   327000        Seattle, WA               10.0   
62638                   237000        Redmond, WA                2.0   
62639                   220000        Seattle, WA               14.0   
62640                   280000  San Francisco, CA                8.0   
62641                   200000      Sunnyvale, CA                0.0   

       yearsatcompany                             tag  basesalary  \
0                 1.5                             NaN    107000.0   
1                 3.0                             NaN         0.0   
2                 0.0                             NaN    155000.0   
3                 5.0                             NaN    157000.0   
4                 3.0                             NaN         0.0   
...               ...                             ...         ...   
62637             1.0  Distributed Systems (Back-End)    155000.0   
62638             2.0                      Full Stack    146900.0   
62639            12.0                      Full Stack    157000.0   
62640             4.0                             iOS    194688.0   
62641             0.0                         ML / AI    155000.0   

       stockgrantvalue    bonus gender otherdetails  cityid  dmaid  rowNumber  \
0              20000.0  10000.0    NaN          NaN    7392  807.0          1   
1                  0.0      0.0    NaN          NaN    7419  807.0          2   
2                  0.0      0.0    NaN          NaN   11527  819.0          3   
3             180000.0  35000.0    NaN          NaN    7472  807.0          7   
4                  0.0      0.0    NaN          NaN    7322  807.0          9   
...                ...      ...    ...          ...     ...    ...        ...   
62637         150000.0  22000.0    NaN          NaN   11527  819.0       1973   
62638          73200.0  16000.0    NaN          NaN   11521  819.0       2037   
62639          25000.0  20000.0    NaN          NaN   11527  819.0       2044   
62640          57000.0  29000.0    NaN          NaN    7419  807.0       2097   
62641          27000.0  19000.0   Male          NaN    7472  807.0       5424   

       Masters_Degree  Bachelors_Degree  Doctorate_Degree  Highschool  \
0                   0                 0                 0           0   
1                   0                 0                 0           0   
2                   0                 0                 0           0   
3                   0                 0                 0           0   
4          

## Penjelasan data yang digunakan

Data yang kami gunakan adalah data mengenai gaji dari perusahaan terbaik dunia. <br>
Tujuan kami menggunakan data ini yaitu untuk mengetahui prospek pekerjaan dan gaji mengenai dunia teknologi yang selaras dengan fakultas STEI mengacu pada perusahaan besar global, sehingga diharapkan dapat memahami terkait gaji dan pekerjaan yang sedang banyak kuotanya. <br>
Sumber data kami adalah https://www.kaggle.com/datasets/jackogozaly/data-science-and-stem-salaries <br>
Data tersebut adalah hasil pengumpulan data oleh kaggle dari https://www.levels.fyi/?compare=Google,Facebook,Microsoft&track=Software%20Engineer . <br>
Data yang kami ambil berisi mengenai informasi tingkat pendidikan, kompensasi (gaji pokok, bonus, hibah saham),\
ras pegawai, dan lainnya.
format data kita adalah **csv** atau dengan nama lain "comma seperated values"\
ukuran filenya adalah 12,544 MB dengan
dimensi dan total elemen dari data kita adalah sebagai berikut :

### Jumlah Baris dan Kolom

In [335]:
print("Jumlah baris:", len(data))
print("Jumlah kolom:", len(data.columns))
    # menampilkan dimensi dari data
    # (baris,kolom)

Jumlah baris: 62642
Jumlah kolom: 29


### Jumlah Elemen Data

In [336]:
print("Total elemen data:", data.size)
    # menampilkan total elemen data

Total elemen data: 1816618


## Penjelasan atribut/kolom

Data kita mengandung kolom-kolom sebagai berikut


In [337]:
print("Atribut pada kolom:\n", data.columns.to_list())

Atribut pada kolom:
 ['timestamp', 'company', 'level', 'title', 'totalyearlycompensation', 'location', 'yearsofexperience', 'yearsatcompany', 'tag', 'basesalary', 'stockgrantvalue', 'bonus', 'gender', 'otherdetails', 'cityid', 'dmaid', 'rowNumber', 'Masters_Degree', 'Bachelors_Degree', 'Doctorate_Degree', 'Highschool', 'Some_College', 'Race_Asian', 'Race_White', 'Race_Two_Or_More', 'Race_Black', 'Race_Hispanic', 'Race', 'Education']


Info dari data yang mencakup isi kolom, indeks kolom, null type atribut kolom, dan tipe dari isi kolom, serta jumlah dari tipe data total pada kolom adalah sebagai berikut

In [338]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 62642 entries, 0 to 62641
Data columns (total 29 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   timestamp                62642 non-null  object 
 1   company                  62637 non-null  object 
 2   level                    62523 non-null  object 
 3   title                    62642 non-null  object 
 4   totalyearlycompensation  62642 non-null  int64  
 5   location                 62642 non-null  object 
 6   yearsofexperience        62642 non-null  float64
 7   yearsatcompany           62642 non-null  float64
 8   tag                      61788 non-null  object 
 9   basesalary               62642 non-null  float64
 10  stockgrantvalue          62642 non-null  float64
 11  bonus                    62642 non-null  float64
 12  gender                   43102 non-null  object 
 13  otherdetails             40137 non-null  object 
 14  cityid                

Tipe data yang terdapat pada tiap kolom bersifat sebagai berikut :
- Categorical - Ordinal (tipe data ordering number) : <br> 
cityid, rowNumber, dan dmaid 

- Categorical - Nominal (label variabel tanpa urutan atau nilai kuantitatif): <br>
company, level, title, location, tag, gender, otherdetails, Race, dan Education

- Quantitative - Continues (data angka yang bersifat kontinu): <br>
timestamp

- Quantitative - Discrete (data bilangan yang dapat dihitung): <br>
totalyearlycompensation, yearsofexperience, yearsatcompany, basesalary, stockgrantvalue, bonus, Masters_Degree, Bachelors_Degree, Doctorate, Highschool, Some_College, Race_Asian, Race_White, Race_Two_Or_More, Race_Black, dan Race_Hispanic


Karakteristik - karakteristik dan definisi yang terdapat pada kolom adalah sebagai berikut :
- timestamp : terdiri dari data angka yang berkaitan dengan tanggal dan jam pengambilan data
- totalyearlycompensation : terdiri dari data angka kompensasi atau gaji tahunan dalam USD
- yearsofexperience : terdiri dari data angka terkait berapa lama pengalaman terjun dalam profesi yang sama dalam tahun
- yearsatcompany : terdiri dari data angka terkait berapa lama menetap pada perusahaan yang sama dalam tahun
- basesalary : terdiri dari data angka terkait gaji pokok minimum sebagai imbalan dalam USD
- stockgrantvalue : terdiri dari data angka terkait bagian saham yang diperoleh dalam USD
- bonus : terdiri dari data angka terkait bonus yang diperoleh dalam USD
- rowNumber : berisi data angka nomor baris dari data kotor kaggle
- Masters_Degree : berisi data untuk mengidentifikasi lulusan s2 atau bukan
- Bachelors_Degree : berisi data angka untuk mengidentifikasi lulusan s1 atau bukan
- Doctorate_Degree : berisi data angka untuk mengidentifikasi lulusan s3 atau bukan
- Highschool : berisi data angka untuk mengidentifikasi siswa SMA
- Some_College : berisi data angka untuk mengidentifikasi termasuk seseorang some college (telah kuliah tetapi mereka tidak menerima gelar apa pun.)
- Race_Asian : berisi data untuk mengidentifikasi ras asia
- Race_White : berisi data untuk mengidentifikasi ras kulit putih
- Race_Two_Or_More : berisi data untuk mengidentifikasi ras campuran
- Race_Black : berisi data untuk mengidentifikasi ras kulit hitam
- Race_Hispanic : berisi data untuk mengidentifikasi ras kulit hispanik

### Menampilkan isi data kategorikal nominal

- company : terdiri dari nama-nama perusahaan sebagai berikut

In [339]:
data["company"].unique()

array(['Oracle', 'eBay', 'Amazon', ..., 'VmWare', 'SalesForce',
       'tableau software'], dtype=object)

- level : terdiri dari tingkat jabatan dalam suatu perusahaan tertentu sebagai berikut

In [340]:
data["level"].unique()

array(['L3', 'SE 2', 'L7', ..., 'L6 (Staff)', 'Mts 1',
       'Bioinformatics Scientist II'], dtype=object)

- title : terdiri dari title jabatan terkait pekerjaan dalam suatu perusahaan sebagai berikut

In [341]:
data["title"].unique()

array(['Product Manager', 'Software Engineer',
       'Software Engineering Manager', 'Data Scientist',
       'Solution Architect', 'Technical Program Manager',
       'Human Resources', 'Product Designer', 'Marketing',
       'Business Analyst', 'Hardware Engineer', 'Sales', 'Recruiter',
       'Mechanical Engineer', 'Management Consultant'], dtype=object)

- location : berisi lokasi tempat suatu perusahaan tertentu sebagai berikut

In [342]:
data["location"].unique()

array(['Redwood City, CA', 'San Francisco, CA', 'Seattle, WA', ...,
       'Pasir Ris, SG, Singapore', 'Richmond, VA, United States',
       'Hilbert, WI'], dtype=object)

- tag : berisi terkait bidang profesi spesifik yang harus dihadapi sebagai berikut

In [343]:
data["tag"].unique()

array([nan, 'Full Stack', 'Distributed Systems (Back-End)', ...,
       'Solution Mgt', 'Software Infra', 'Industry Cloud'], dtype=object)

- gender : berisi jenis kelamin seorang pengisi data

In [344]:
data["gender"].unique()

array([nan, 'Male', 'Female', 'Other', 'Title: Senior Software Engineer'],
      dtype=object)

- otherdetails : berisi keterangan tambahan terkait ketentuan mauapun kesepakatan pekerjaan

In [345]:
data["otherdetails"].unique()

array([nan, 'Tons of overtime', 'PhD', ...,
       '40k cash split over 2 years. 75k rsus over 4 years. Bonus is median 10% of base per yr.',
       '65k of stocks at 5/15/40/40 vesting schedule per year. First job out of college, only internship experience prior',
       'industry hire'], dtype=object)

- cityid : berisi data angka terkait kode kota suatu perusahaan berada

In [346]:
data["cityid"].unique()

array([ 7392,  7419, 11527, ...,  3799, 16802, 28230], dtype=int64)

- dmaid : berisi data kode designated market area dari perusahaan

In [347]:
data["dmaid"].unique()

array([807., 819.,   0., 501., 512., 862., 820., 803., 511., 635., 506.,
       825., 622., 508., 519., 751., 652., 504., 527., 617., 613., 602.,
       532., 753., 752., 623., 556., 669., 628., 566., 515., 560., 528.,
       534., 524., 609., 526., 521., 648., 770., 517., 518., 535., 505.,
       641., 790., 539., 670., 533., 523., 618., 637., 616., 855., 510.,
       563., 757., 659., 625., 691., 705., 576., 811., 679., 541., 544.,
       828., 789., 821., 588., 611., 716., 801., 561., 682., 584., 671.,
        nan, 564., 502., 640., 678., 581., 642., 514., 598., 643., 567.,
       839., 746., 500., 675., 619., 548., 592., 650., 529., 536., 551.,
       573., 610., 766., 702., 724., 547., 604., 656., 513., 649., 503.,
       687., 530., 686., 555., 571., 538., 693., 651., 744., 634., 525.,
       546., 754., 630., 624., 866., 554., 600., 565., 542., 550., 540.,
       632., 734., 520., 881., 509., 522., 612., 531., 804., 577., 759.,
       516., 673., 574., 537., 760., 698., 658.])

- Race : berisi data ras pegawai

In [348]:
data["Race"].unique()

array([nan, 'White', 'Asian', 'Black', 'Two Or More', 'Hispanic'],
      dtype=object)

- Education: berisi data pendidikan terakhir

In [349]:
data["Education"].unique()

array([nan, 'PhD', "Master's Degree", "Bachelor's Degree", 'Some College',
       'Highschool'], dtype=object)

### Menampilkan range data numerik

- Pengalaman pekerjaan, range datanya :

In [350]:
min_year_exp = data["yearsofexperience"].min()
max_year_exp = data["yearsofexperience"].max()
print ("Range pengalaman pekerjaan : " + str(min_year_exp) + " - " + str(max_year_exp) + " tahun")


Range pengalaman pekerjaan : 0.0 - 69.0 tahun


- Pengalaman dalam perusahaan yang sama, range datanya :

In [351]:
min_year_comp = data["yearsatcompany"].min()
max_year_comp = data["yearsatcompany"].max()
print ("Range pengalaman bekerja pada suatu perusahaan: " + str(min_year_comp) + " - " + str(max_year_comp) + " tahun")


Range pengalaman bekerja pada suatu perusahaan: 0.0 - 69.0 tahun


- Kompensasi/gaji per tahun, range datanya :

In [352]:
min_year_fee = data["totalyearlycompensation"].min()
max_year_fee = data["totalyearlycompensation"].max()
print ("Range kompensasi per tahun : " + str(min_year_fee) + " - " + str(max_year_fee) + " (dalam USD)")


Range kompensasi per tahun : 10000 - 4980000 (dalam USD)


- Gaji pokok : Gaji minimal, range datanya :

In [353]:
min_base_salary = data["basesalary"].min()
max_base_salary = data["basesalary"].max()
print ("Range gaji pokok per tahun: " + str(min_base_salary) + " - " + str(max_base_salary) + " (dalam USD)")


Range gaji pokok per tahun: 0.0 - 1659870.0 (dalam USD)


- Pemberian saham, range datanya :

In [354]:
min_stock_grant = data["stockgrantvalue"].min()
max_stock_grant = data["stockgrantvalue"].max()
print ("Range pemberian saham per tahun: " + str(min_stock_grant) + " - " + str(max_stock_grant) + " (dalam USD)")


Range pemberian saham per tahun: 0.0 - 2800000.0 (dalam USD)


- Pemberian bonus, range datanya :

In [355]:
min_bonus = data["bonus"].min()
max_bonus = data["bonus"].max()
print ("Range pemberian bonus per tahun: " + str(min_bonus) + " - " + str(max_bonus) + " (dalam USD)")


Range pemberian bonus per tahun: 0.0 - 1000000.0 (dalam USD)


- Sebaran Data Ras Pegawai

In [356]:
print("Data ras terdiri dari: ",data["Race"].unique().tolist())
print("Dengan rincian: ")
print("Jumlah pekerja ras asia:", data["Race_Asian"].sum(),"orang")
print("Jumlah pekerja ras kulit putih:", data["Race_White"].sum(),"orang")
print("Jumlah pekerja ras kulit hitam:", data["Race_Black"].sum(),"orang")
print("Jumlah pekerja ras hispanic:", data["Race_Hispanic"].sum(),"orang")
print("Jumlah pekerja ras campuran:", data["Race_Two_Or_More"].sum(),"orang")

Data ras terdiri dari:  [nan, 'White', 'Asian', 'Black', 'Two Or More', 'Hispanic']
Dengan rincian: 
Jumlah pekerja ras asia: 11772 orang
Jumlah pekerja ras kulit putih: 8032 orang
Jumlah pekerja ras kulit hitam: 690 orang
Jumlah pekerja ras hispanic: 1130 orang
Jumlah pekerja ras campuran: 804 orang


- Sebaran Data Pendidikan Pegawai

In [357]:
print("Data pendidikan terdiri dari:\n",data["Education"].unique().tolist())
print("Dengan rincian: ")
print("Lulusan SMA:",data["Highschool"].sum(),"orang")
print("Some collage:",data["Some_College"].sum(),"orang")
print("Lulusan S1:",data["Bachelors_Degree"].sum(),"orang")
print("Lulusan S2:",data["Masters_Degree"].sum(),"orang")
print("Lulusan S3:",data["Doctorate_Degree"].sum(),"orang")

Data pendidikan terdiri dari:
 [nan, 'PhD', "Master's Degree", "Bachelor's Degree", 'Some College', 'Highschool']
Dengan rincian: 
Lulusan SMA: 320 orang
Some collage: 355 orang
Lulusan S1: 12605 orang
Lulusan S2: 15391 orang
Lulusan S3: 1803 orang


## Presentase data kosong


In [358]:
x = (data.isna().sum().sum()/data.size)*100
print("Presentase data kosong adalah "+"{:.2f}%".format(x))

Presentase data kosong adalah 6.36%


### Mengecek data kosong


In [359]:
data.isna().sum()
    # mengecek kolom yang mengandung data kosong
    # kolom yang mengandung data kosong dalam data kami adalah company, level, dmaid, tag, gender,otherdetails, race, dan education 

timestamp                      0
company                        5
level                        119
title                          0
totalyearlycompensation        0
location                       0
yearsofexperience              0
yearsatcompany                 0
tag                          854
basesalary                     0
stockgrantvalue                0
bonus                          0
gender                     19540
otherdetails               22505
cityid                         0
dmaid                          2
rowNumber                      0
Masters_Degree                 0
Bachelors_Degree               0
Doctorate_Degree               0
Highschool                     0
Some_College                   0
Race_Asian                     0
Race_White                     0
Race_Two_Or_More               0
Race_Black                     0
Race_Hispanic                  0
Race                       40215
Education                  32272
dtype: int64

### Persentase data kosong pada kolom 


In [360]:
list_NA_data = ["company","level","dmaid","tag","gender","otherdetails","Race","Education"]
for i in list_NA_data:
    print("Persentase data kosong kolom",i,": {:.3f}%".format((data[i].isna().sum()/data[i].size)*100))
# Persentase darta kosong pada masing-masing kolom

Persentase data kosong kolom company : 0.008%
Persentase data kosong kolom level : 0.190%
Persentase data kosong kolom dmaid : 0.003%
Persentase data kosong kolom tag : 1.363%
Persentase data kosong kolom gender : 31.193%
Persentase data kosong kolom otherdetails : 35.926%
Persentase data kosong kolom Race : 64.198%
Persentase data kosong kolom Education : 51.518%


### Presentase data duplikat


Tidak ada data duplikat pada data kami karena tiap orang mungkin memiliki kesamaan pada atribut dalam data kami, yakni kesamaan pekerjaan, perusahaan, gender, pendidikan, ras, dll.

### Presentase data salah

In [361]:
# still thinking

In [362]:
import datetime
data['Year'] = pd.DatetimeIndex(data['timestamp']).year
data['Month'] = pd.DatetimeIndex(data['timestamp']).month
data['Day'] = pd.DatetimeIndex(data['timestamp']).day
data['Date'] = pd.DatetimeIndex(data['timestamp']).date
import calendar
data['Month'] = data['Month'].apply(lambda x: calendar.month_abbr[x])


## Data Checking dan Data Cleaning
kolom data kosong dalam data kami adalah company, level, dmaid, tag, gender,otherdetails, race, dan education 



### Prosentase Kekosongan Data Besar
- ### <b>Data Education</b> 

#### Data Checking Education<br> 
Checking yang kami lakukan:
- Mengidentifikasi apakah maksud data dalam spesifik lulusan pendidikan adalah 1 untuk True dan 0 untuk false
- Untuk data kosong pada education apakah terdapat data pendidikan pada data spesifik lulusan pendidikan, jika terdapat datanya
maka data kosong diisi data pendidikan terakhir pada data spesifik tersebut

In [363]:
data.loc[(data["Masters_Degree"] != 0) & (data["Masters_Degree"] != 1)]
# Ambil sample data lulusan S2 dan cek apakah ada angka selain 1 dan 0

Empty DataFrame
Columns: [timestamp, company, level, title, totalyearlycompensation, location, yearsofexperience, yearsatcompany, tag, basesalary, stockgrantvalue, bonus, gender, otherdetails, cityid, dmaid, rowNumber, Masters_Degree, Bachelors_Degree, Doctorate_Degree, Highschool, Some_College, Race_Asian, Race_White, Race_Two_Or_More, Race_Black, Race_Hispanic, Race, Education, Year, Month, Day, Date]
Index: []

Tidak ada data yang ditampilkan, artinya data asumsi awal benar berkaitan bahwa nilai 1 adalah true dan 0 adalah false
Maka dari itu agar tidak membingungkan user selanjutnya isi dari kolom tersebut sebaiknya semua data berkaitan dengan pendidikan diganti dengan boolean True or False seperti berikut 

#### Data Cleaning Education

Mengganti nilai tiap kolom pendidikan dari 1 menjadi True dan 0 menjadi False

In [364]:
# list_education = ["Masters_Degree","Bachelors_Degree","Doctorate_Degree","Highschool","Some_College"]
# # Mengganti data 1 menjadi true dan 0 menjadi false
# for i in list_NA_data:
#     data[list_education] = data[list_education].replace([1],True)
#     data[list_education] = data[list_education].replace([0],False)


In [365]:

list_education = ["Masters_Degree","Bachelors_Degree","Doctorate_Degree","Highschool","Some_College"]
# Menampilkan isi dari kolom data setelah diganti
for i in list_education:
    print(data[i].unique())
# Atribut dari masing-masing kolom adalah True dan False dan tidak ada data selain itu 
# berarti asumsi awal berkaitan definisi nilai 1 dan 0 adalah benar

[0 1]
[0 1]
[0 1]
[0 1]
[0 1]


Namun dalam data pendidikan ini terdapat keanehan yaitu misalnya jika lulusan S2 True maka lulusan S1 False, padahal untuk dapat mencapai pendidikan yang lebih tinggi harus menyelesaikan pendidikan yang lebih rendah dulu, contoh untuk dapat kuliah S2, maka harus mendapat surat rekomendasi lulusan S1, dan untuk dapat kuliah S1 maka harus lulus SMA

In [366]:
# Ambil sampel
sample = data.loc[(data["Masters_Degree"]==1) & (data["company"]=="Amazon")]
sample[["Masters_Degree","Bachelors_Degree","Doctorate_Degree","Highschool","Some_College"]]

Masters_Degree  Bachelors_Degree  Doctorate_Degree  Highschool  \
12506               1                 0                 1           0   
12512               1                 0                 0           0   
12528               1                 0                 0           0   
12551               1                 0                 0           0   
12558               1                 0                 0           0   
...               ...               ...               ...         ...   
61683               1                 0                 0           0   
61806               1                 0                 0           0   
61809               1                 0                 0           0   
61810               1                 0                 0           0   
61817               1                 0                 0           0   

       Some_College  
12506             0  
12512             0  
12528             0  
12551             0  
12558             0  
...             ...  
61683             0  
61806             0  
61809             0  
61810             0  
61817             0  

[2080 rows x 5 columns]

In [367]:
# Cari erornya 
# Kalau dia udah s1 brarti udah sma dst
# replace data untuk kasus itu

high_school = data.loc[(data["Masters_Degree"]==1) | (data["Bachelors_Degree"]==1) | (data["Doctorate_Degree"]==1) | (data["Some_College"]==1)]
bachelor = data.loc[(data["Masters_Degree"]==1) | (data["Doctorate_Degree"]==1)]
master = data.loc[data["Doctorate_Degree"]==1]
# high_school["Highschool"] = high_school["Highschool"].replace(([0], 1))
# bachelor["Bachelors_Degree"] = bachelor["Bachelors_Degree"].replace(([0], 1))
# master["Masters_Degree"] = master["Masters_Degree"].replace(([0], 1))
high_school["Highschool"].fillna(1, inplace=True)
# high_school["Highschool"] = high_school["Highschool"].replace([0], 1)
# data["Bachelors_Degree"] = bachelor["Bachelors_Degree"].replace([0], 1)
# data["Masters_Degree"] = master["Masters_Degree"].replace([0], 1)
# data[["Masters_Degree","Bachelors_Degree","Doctorate_Degree","Highschool","Some_College"]]
# print("Data rincian pendidikan menjadi: ")
print("Data pendidikan terdiri dari:\n",data["Education"].unique().tolist())
print("Dengan rincian: ")
print("Lulusan SMA:",data["Highschool"].sum(),"orang")
print("Some collage:",data["Some_College"].sum(),"orang")
print("Lulusan S1:",data["Bachelors_Degree"].sum(),"orang")
print("Lulusan S2:",data["Masters_Degree"].sum(),"orang")
print("Lulusan S3:",data["Doctorate_Degree"].sum(),"orang")

    

Data pendidikan terdiri dari:
 [nan, 'PhD', "Master's Degree", "Bachelor's Degree", 'Some College', 'Highschool']
Dengan rincian: 
Lulusan SMA: 320 orang
Some collage: 355 orang
Lulusan S1: 12605 orang
Lulusan S2: 15391 orang
Lulusan S3: 1803 orang


- ### <b>Data Race</b> 

#### Data Checking Ras
Checking yang kami lakukan :
- Mengidentifikasi apakah angka yang dimaksud dalam ras spesifik adalah 1 untuk True dan 0 untuk false
- Untuk data kosong pada ras apakah terdapat data ras spesifik, jika terdapat data aras spesifik, maka kekosongan data akan diisi dengan ras spesifik

In [368]:
data.loc[(data["Race_Asian"] != 0) & (data["Race_Asian"] != 1)]
# Ambil sample data ras spesifik dan cek apakah ada angka selain 1 dan 0

Empty DataFrame
Columns: [timestamp, company, level, title, totalyearlycompensation, location, yearsofexperience, yearsatcompany, tag, basesalary, stockgrantvalue, bonus, gender, otherdetails, cityid, dmaid, rowNumber, Masters_Degree, Bachelors_Degree, Doctorate_Degree, Highschool, Some_College, Race_Asian, Race_White, Race_Two_Or_More, Race_Black, Race_Hispanic, Race, Education, Year, Month, Day, Date]
Index: []

Tidak ada data yang ditampilkan, artinya data asumsi awal benar berkaitan bahwa nilai 1 adalah true dan 0 adalah false. Maka selanjutnya untuk cleansing data kami mengecek apakah ada nilai pada Race_Asian/Race_White dan ras lainnya sementara pada data nominal ras kosong


In [369]:
data.loc[((data["Race_Asian"]==1) | (data["Race_White"]==1)|(data["Race_Black"]==1)|(data["Race_Hispanic"]==1)|(data["Race_Two_Or_More"]==1)) & data["Race"]=="NA"]
# Mengecek : apakah ada nilai pada Race_Asian/Race_White dan ras lainnya sementara pada data nominal ras kosong

Empty DataFrame
Columns: [timestamp, company, level, title, totalyearlycompensation, location, yearsofexperience, yearsatcompany, tag, basesalary, stockgrantvalue, bonus, gender, otherdetails, cityid, dmaid, rowNumber, Masters_Degree, Bachelors_Degree, Doctorate_Degree, Highschool, Some_College, Race_Asian, Race_White, Race_Two_Or_More, Race_Black, Race_Hispanic, Race, Education, Year, Month, Day, Date]
Index: []

#### Data Cleaning Ras


Karena kekosongan data murni akibat tidak adanya data, kemudian dengan pertimbangan bahwa persentase kekosongan data tersebut cukup besar 64.20% maka tidak bisa semena-mena dihapus, kami mengambil tindakan untuk dikategorikan ke dalam Unknown Race atau Ras tidak diketahui

In [370]:
data["Race"].fillna("Unknown Race", inplace=True)
# Mengganti data kosong menjadi unknown race

In [371]:
print("Persentase data kosong kolom Ras"+": {:.2f}%".format((data["Race"].isna().sum()/data["Race"].size)*100))
# Validasi kekosongan data setelah cleaning

Persentase data kosong kolom Ras: 0.00%


- ### <b>Data otherdetails</b> 
#### Data Cleaning otherdetails
Data pada otherdetails hanya berkaitan dengan detail tambahan terkait ketentuan dan kesepakatan dalam pekerjaan
sehingga atas pertimbangan tersebut dan pertimbangan kekosongan data 35.93% yang cukup besar maka kami mengambil tindakan untuk merubah data kosong menjadi Nothing details atau tidak ada detail tambahan

In [372]:
data["otherdetails"].fillna("Nothing details", inplace=True)
# Mengganti data kosong menjadi nothing details

In [373]:
print("Persentase data kosong kolom otherdetails"+": {:.3f}%".format((data["otherdetails"].isna().sum()/data["otherdetails"].size)*100))
# Validasi kekosongan data setelah cleaning

Persentase data kosong kolom otherdetails: 0.000%


- ### <b>Data gender</b> 
#### Data Cleaning gender
Data pada gender hanya berkaitan dengan jenis kelamin laki-laki atau perempuan. Maka dengan pertimbangan tersebut dan pertimbangan bahwa kekosongan data gender 31.19% yang cukup besar maka kami mengambil tindakan untuk merubah data kosong menjadi don't want to be known atau tidak ingin diketahui karena privasi pribadi

In [374]:
data["gender"].fillna("Don't want to be known", inplace=True)
# Mengganti data kosong menjadi don't want to be known

In [375]:
print("Persentase data kosong kolom gender"+": {:.3f}%".format((data["gender"].isna().sum()/data["gender"].size)*100))
# Validasi kekosongan data setelah cleaning

Persentase data kosong kolom gender: 0.000%


### Prosentase Kekosongan Data Kecil
- ### <b>Data company, level, dmaid, dan tag</b> 
#### Data Cleaning pada data dengan Prosentase Kekosongan Data Kecil
Data berkaitan company, level, dmaid, dan tag yang prosentase null data hanya 0.003-1.363% dinilai cukup kecil dan tidak akan terlalu merubah data jika dihapus, maka kami mengambil tindakan untuk menghapus data yang prosentase null datanya kecil

In [376]:
data.dropna(subset=['company','level','dmaid','tag'],inplace=True)
# Menghapus data yang kosong pada kolom company, level, dmaid, dan tag

In [377]:
list_NA_data = ["company","level","dmaid","tag","gender","otherdetails","Race","Education"]
for i in list_NA_data:
    print("Persentase data kosong kolom",i,": {:.3f}%".format((data[i].isna().sum()/data[i].size)*100))
# Menampilkan validasi hasil cleansing data

Persentase data kosong kolom company : 0.000%
Persentase data kosong kolom level : 0.000%
Persentase data kosong kolom dmaid : 0.000%
Persentase data kosong kolom tag : 0.000%
Persentase data kosong kolom gender : 0.000%
Persentase data kosong kolom otherdetails : 0.000%
Persentase data kosong kolom Race : 0.000%
Persentase data kosong kolom Education : 50.901%


### Wrong Data
- ### <b>Data totalyearlycompensation dan basesalary</b> 
Data gaji pokok adalah data gaji minimum yang didapatkan pegawai tidak mungkin bernilai 0 apalagi perusahaan terkait termasuk perusahaan yang besar, namun untuk pengalaman bekerja, pengalaman pada perusahaan, bonus, dan pembagian saham masih mungkin bernilai 0 karena bisa jadi baru mulai bekerja dan merupakan pegawai baru
#### Data Cleaning pada data wrong data
Data berkaitan totalyearlycompensation dan basesalary yang bernilai 0 akan dihapus karena termasuk wrong data

In [378]:
for i in data["basesalary"].index:
    if data.loc[i, "basesalary"] <= 0:
        data.drop(i, inplace=True)
# Menghapus semua data yang bernilai 0 dan negatif (jika ada)

In [379]:
min_base_salary = data["basesalary"].min()
max_base_salary = data["basesalary"].max()
print ("Range gaji pokok per tahun: " + str(min_base_salary) + " - " + str(max_base_salary) + " (dalam USD)")
# validasi setelah cleansing data

Range gaji pokok per tahun: 1000.0 - 1620000.0 (dalam USD)


## Mengambil sample 10000 baris di tengah data terurut yang sudah dibersihkan
Kami terlebih dahulu mengurutkan data berdasarkan total gaji yang diterima, pengalaman kerja, dan pengalaman pada perusahaan kemudian kami mengambil data tengah dengan jumlah 10000 baris sebagai sampel data mengingat bahwa jumlah baris pada data asli kami sangat banyak yaitu 660000 dan pada banyak perusahaan yang berbeda-beda dan kami mengambil nilai tengahnya agar tidak terjadi simpangan data yang juga semakin jauh


In [380]:
samples = data.sort_values(by=["totalyearlycompensation","yearsofexperience","yearsatcompany"], ascending=True)
samples = samples.iloc[25000:35000].copy()

In [386]:
samples

timestamp           company                       level  \
29869  9/14/2020 11:24:16  Juniper Networks         Software Engineer 4   
35280  11/15/2020 8:40:51          Facebook                          E4   
49433  4/23/2021 21:59:41              Citi              Vice President   
6418   5/11/2019 11:20:51            Amazon                      SDE II   
30779  9/23/2020 17:59:26            PayPal                          L2   
...                   ...               ...                         ...   
39296  1/13/2021 10:51:51          Citibank                         C14   
15616  1/26/2020 10:41:41              eBay                       MTS 1   
22801  6/19/2020 10:00:29             Tesla             Senior Engineer   
52958  5/31/2021 14:52:03             Apple                        ICT4   
49069  4/21/2021 12:07:27      Rally Health  Senior Software Engineer 2   

                   title  totalyearlycompensation  \
29869  Software Engineer                   170000   
35280  Software Engineer                   170000   
49433  Software Engineer                   170000   
6418   Software Engineer                   170000   
30779   Product Designer                   170000   
...                  ...                      ...   
39296     Data Scientist                   209000   
15616  Software Engineer                   209000   
22801  Software Engineer                   209000   
52958  Hardware Engineer                   209000   
49069  Software Engineer                   209000   

                          location  yearsofexperience  yearsatcompany  \
29869                Sunnyvale, CA                8.0             1.0   
35280   London, EN, United Kingdom                8.0             1.0   
49433                   Dallas, TX                8.0             1.0   
6418         Vancouver, BC, Canada                8.0             2.0   
30779                 San Jose, CA                8.0             2.0   
...                            ...                ...             ...   
39296                  Atlanta, GA               11.0            11.0   
15616                 Bellevue, WA               12.0             1.0   
22801           Salt Lake City, UT               12.0             1.0   
52958  Swindon, EN, United Kingdom               12.0             6.0   
49069              Minneapolis, MN               13.0             2.0   

                                  tag  basesalary  stockgrantvalue    bonus  \
29869                      Networking    140000.0          10000.0  20000.0   
35280                      Full Stack    103000.0          50000.0  17000.0   
49433                        Security    162000.0              0.0   8000.0   
6418       API Development (Back-End)    120000.0          30000.0  10000.0   
30779            User Experience (UX)    140000.0          20000.0  10000.0   
...                               ...         ...              ...      ...   
39296                            Risk    167000.0              0.0  42000.0   
15616                      Full Stack    172000.0          20000.0  17000.0   
22801  Distributed Systems (Back-End)    135000.0          44000.0  30000.0   
52958                          Analog    114000.0          80000.0  15000.0   
49069                            Data    172000.0          20000.0  17000.0   

                       gender  \
29869  Don't want to be known   
35280                    Male   
49433                    Male   
6418                     Male   
30779  Don't want to be known   
...                       ...   
39296                  Female   
15616                    Male   
22801                    Male   
52958  Don't want to be known   
49069                    Male   

                                            otherdetails  cityid  dmaid  \
29869                                    Nothing details    7472  807.0   
35280  Title: Software Engineer, Race: Asian, Academi...   12008    0.0   
49433  10k relo bonus (employe

# Statistik pada sample


In [382]:
samples[["totalyearlycompensation", "yearsofexperience", "yearsatcompany","basesalary", "stockgrantvalue", "bonus", "Masters_Degree", "Bachelors_Degree", "Doctorate_Degree", "Highschool", "Some_College", "Race_Asian", "Race_White", "Race_Black", "Race_Hispanic", "Race_Two_Or_More"]].describe(percentiles = [.1,.25,.50,.75,.90]).applymap('{:,.4f}'.format)
# Mengambil deskripsi dari data sample dengan format 4 angka di belakang koma

totalyearlycompensation yearsofexperience yearsatcompany    basesalary  \
count             10,000.0000       10,000.0000    10,000.0000   10,000.0000   
mean             188,709.5500            6.3452         2.4962  143,541.5642   
std               10,889.8357            5.4013         3.1295   18,293.7071   
min              170,000.0000            0.0000         0.0000   55,000.0000   
10%              175,000.0000            1.0000         0.0000  121,000.0000   
25%              180,000.0000            2.0000         0.0000  132,000.0000   
50%              189,000.0000            5.0000         2.0000  143,000.0000   
75%              200,000.0000            9.0000         3.0000  155,000.0000   
90%              204,000.0000           14.0000         6.0000  166,000.0000   
max              209,000.0000           41.0000        25.0000  208,000.0000   

      stockgrantvalue         bonus Masters_Degree Bachelors_Degree  \
count     10,000.0000   10,000.0000    10,000.0000      10,000.0000   
mean      27,987.6998   17,413.1409         0.2797           0.1819   
std       21,316.9168   12,931.5980         0.4489           0.3858   
min            0.0000        0.0000         0.0000           0.0000   
10%            0.0000        0.0000         0.0000           0.0000   
25%       13,000.0000   10,000.0000         0.0000           0.0000   
50%       25,000.0000   17,000.0000         0.0000           0.0000   
75%       40,000.0000   24,000.0000         1.0000           0.0000   
90%       54,000.0000   32,000.0000         1.0000           1.0000   
max      450,000.0000  110,000.0000         1.0000           1.0000   

      Doctorate_Degree   Highschool Some_College   Race_Asian   Race_White  \
count      10,000.0000  10,000.0000  10,000.0000  10,000.0000  10,000.0000   
mean            0.0269       0.0042       0.0047       0.1919       0.1234   
std             0.1618       0.0647       0.0684       0.3938       0.3289   
min             0.0000       0.0000       0.0000       0.0000       0.0000   
10%             0.0000       0.0000       0.0000       0.0000       0.0000   
25%             0.0000       0.0000       0.0000       0.0000       0.0000   
50%             0.0000       0.0000       0.0000       0.0000       0.0000   
75%             0.0000       0.0000       0.0000       0.0000       0.0000   
90%             0.0000       0.0000       0.0000       1.0000       1.0000   
max             1.0000       1.0000       1.0000       1.0000       1.0000   

        Race_Black Race_Hispanic Race_Two_Or_More  
count  10,000.0000   10,000.0000      10,000.0000  
mean        0.0111        0.0166           0.0135  
std         0.1048        0.1278           0.1154  
min         0.0000        0.0000           0.0000  
10%         0.0000        0.0000           0.0000  
25%         0.0000        0.0000           0.0000  
50%         0.0000        0.0000           0.0000  
75%         0.0000        0.0000           0.0000  
90%         0.0000        0.0000           0.0000  
max         1.0000        1.0000           1.0000

## Visualisasi Data



In [383]:
print(data[data["Year"] == 2020].groupby(["Month",""])["s"].sum().unstack())
data6 = data[data["Year"] == 2013].groupby(["MonthNumber","Country"])["Sales"].sum().unstack()

data6.plot(kind="line",title="Sum of Sales per Month per Country 2013",xticks=[9,10,11,12])
plt.show()

In [384]:
# data2 = data[data["Year"] == 2014].groupby("Product")[["Sales","Profit"]].sum()

# data2.plot(kind="line",title="Sales and Profit vs Product 2014")
# plt.show()